### PR TITLE
config service

### PR DESCRIPTION
**Pull request by jjalan**
## Description
Added config service to get env variables.

## Database schema changes
`None`

### Automated test cases added
- ConfigService should return the correct environment
- ConfigService should return correct env variable value
- ConfigService should throw MissingKeyError if the key is not found in the config
- ConfigService should throw a ValueTypeMismatchError if type of value does not match

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -9,8 +9,9 @@
   "main": "src/main.ts",
   "scripts": {
     "build": "gulp",
-    "test": "exit 0",
+    "test": "NODE_ENV=testing mocha --exit -r ts-node/register --project tsconfig.json test/**/*.spec.ts",
```
Comment (jjalan) [inline package.json:12]: NODE_CONFIG_ENV (I think) should be testing and NODE_ENV as production.

This instructs node to compile libs in production mode but use testing config.

File: src/modules/config/config-service.ts [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+import {
+  ConfigType,
+  Environment,
+  MissingKeyError,
+  ValueTypeMismatchError,
+} from './types';
+
+export default class ConfigService {
+  public static getEnvironment(): Environment {
+    return config.get('node.env');
```
Comment (jjalan) [inline src/modules/config/config-service.ts:11]: Please read and map to enum. If it cannot be mapped to enu, it should throw

File: src/modules/config/config-service.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  }
+
+  public static getBoolValue(key: string): boolean {
+    // This method should throw MissingKeyError if key is not found in config
+    // This method should throw ValueTypeMismatchError if key is found but the
+    // there is a value type mistmatch
+    // Otherwise return the value
+    const expectedValueType = ConfigType.BOOLEAN;
+    const value = config.get(key);
```
Comment (jjalan) [inline src/modules/config/config-service.ts:20]: I see this code repeated below.

Possible to refactor?

File: config/custom-environment-variables.json [hunk 4]
```diff
@@ -1 +1,5 @@
-{}
+{
+  "node": {
+    "env": "NODE_ENV"
```
[the remark this answers is not in the document]
Comment (jjalan) [reply config/custom-environment-variables.json:3]: Can I request to reconfirm this by reading config npm module documentation?

**Review by jjalan — COMMENTED** [review]
We should also add a file `qa.json` and `beta.json`.

File: config/custom-environment-variables.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (jjalan) [reply config/custom-environment-variables.json:3]: We should add a spec to read a sample value from each environment.

File: src/modules/config/config-service.ts [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import config from 'config';
+import { Environment, MissingKeyError, ValueTypeMismatchError } from './types';
+
+export enum ConfigType {
```
Comment (jjalan) [inline src/modules/config/config-service.ts:4]: Please move it to internals/config-type.ts

Service should not contain details or should contain very minimal detail so reader of this file may not know much about implementation details please.

File: src/modules/config/config-service.ts [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+  STRING = 'string',
+}
+
+export default class ConfigService {
+  public static getEnvironment(): Environment {
+    const env: Environment = config.get('node.config.env');
+    if (Object.values(Environment).includes(env)) {
+      return env;
+    }
+    throw new Error(`Environment ${env} is not supported`);
```
Comment (jjalan) [inline src/modules/config/config-service.ts:16]: We should create `EnvironmentNotSupportedError`

File: src/modules/config/config-service.ts [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    return this.getEnvValue<number>(expectedValueType, key);
+  }
+
+  public static getStringValue(key: string): string {
+    const expectedValueType = ConfigType.STRING;
+    return this.getEnvValue<string>(expectedValueType, key);
+  }
+
+  private static getEnvValue<T>(expectedValueType: ConfigType, key: string): T {
+    // This method should throw MissingKeyError if key is not found in config
```
Comment (jjalan) [inline src/modules/config/config-service.ts:35]: Remove comments, code is readable.

File: test/spec/config/config.spec.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+import { assert } from 'chai';
+import { Environment } from '../../../src/modules/config/types';
+import ConfigService from '../../../src/modules/config/config-service';
+
+describe('ConfigService', () => {
```
Comment (jjalan) [inline test/spec/config/config.spec.ts:6]: consider adding spec for all environments please.

Comment (akashlakhera) [reply test/spec/config/config.spec.ts:6]: It's very hard to change the config at runtime and make the app throw an `EnvironmentNotSupportedError` error.

**Review by jjalan — COMMENTED** [review]
LGTM

**Pull request merged**